scrappy/scrappy.pyx

```diff
@@ -0,0 +1,740 @@
+        sp = s - sp
+        del s
+
+        # Build the fine structure image :
+        if fsmode == 'convolve':
+            f = convolve(cleanarr, psfk)
+        elif fsmode == 'median':
+            if sepmed:
+                f = sepmedfilt5(cleanarr)
+            else:
+                f = medfilt3(cleanarr)
+        else:
+            raise ValueError('Please choose a valid fine structure mode.')
+
+        if sepmed:
+            m7 = sepmedfilt9(f)
+        else:
+            m7 = medfilt7(f)
+
+        f = (f - m7) / noise
+        # Clip f as we will divide by f. Similar to the IRAF version.
+        f[f < 0.01] = 0.01
+
+        del m7
+        del noise
+
+        # Find the candidate cosmic rays
+        goodpix = np.logical_not(mask)
+        cosmics = np.logical_and(sp > sigclip, goodpix)
+        # Note the sp/f and not lplus/f due to the f = f/noise above.
+        cosmics = np.logical_and(cosmics, (sp / f) > objlim)
+        del f
+
+        # What follows is a special treatment for neighbors, with more relaxed
+        # constraints.
+        # We grow these cosmics a first time to determine the immediate
+        # neighborhood.
+        cosmics = dilate3(cosmics)
+        cosmics = np.logical_and(cosmics, goodpix)
+        # From this grown set, we keep those that have sp > sigmalim
+        cosmics = np.logical_and(sp > sigclip, cosmics)
+
+        # Now we repeat this procedure, but lower the detection limit to siglow
+        cosmics = dilate3(cosmics)
+        cosmics = np.logical_and(cosmics, goodpix)
+
+        del goodpix
+        cosmics = np.logical_and(sp > sigcliplow, cosmics)
+        del sp
+
+        # Our CR counter
+        numcr = cosmics.sum()
+
+        # Update the crmask with the cosmics we have found
+        crmask[:, :] = np.logical_or(crmask, cosmics)[:, :]
+        del cosmics
+        if verbose:
+            print("{} cosmic pixels this iteration".format(numcr))
+
+        # If we didn't find anything, we're done.
+        if numcr == 0:
+            break
+
+        # otherwise clean the image and iterate
+        if cleantype == 'median':
+            # Unmasked median filter
+            cleanarr[crmask] = m5[crmask]
+            del m5
+        # Masked mean filter
+        elif cleantype == 'meanmask':
+            clean_meanmask(cleanarr, crmask, mask, nx, ny, backgroundlevel)
+        # Masked median filter
+        elif cleantype == 'medmask':
+            clean_medmask(cleanarr, crmask, mask, nx, ny, backgroundlevel)
+        # Inverse distance weighted interpolation
+        elif cleantype == 'idw':
+            clean_idwinterp(cleanarr, crmask, mask, nx, ny, backgroundlevel)
+        else:
+            raise ValueError("""cleantype must be one of the following values:
+                            [median, meanmask, medmask, idw]""")
+
+    return (crmask.astype(np.bool), cleanarr)
+
+
+def updatemask(np.ndarray[np.float32_t, ndim=2, mode='c', cast=True] data,
```

Comment:
`updatemask` -> `update_mask`

(I think that's a PEP8 rule or at least common in Python, plus consistent with `clean_meanmask` below)